Updated initial VPN support

shadeofblue opened the pull request:
this is a rework of https://github.com/golemfactory/yapapi/pull/444/files

this functionality requires yagna `0.8.0-rc4` on both the requestor and provider and additionally, ya-runtime-vm `pre-rel-v0.2.9` on the provider's end

provider:
```
curl -sSf https://join.golem.network/as-provider | YA_INSTALLER_CORE=pre-rel-v0.8.0-rc4 YA_INSTALLER_VM=pre-rel-v0.2.9 bash -
```

requestor:
```
curl -sSf https://join.golem.network/as-requestor | YA_INSTALLER_CORE=pre-rel-v0.8.0-rc4 bash -
```

johny-b commented on examples/ssh/ssh.py:
```diff
@@ -0,0 +1,158 @@
+        budget=1.0,
+        subnet_tag=subnet_tag,
+        driver=driver,
+        network=network,
```
I think we have a naming problem here and I'm not sure how to solve it.
There is `Golem.network` but also `golem.create_network` and those are totally different networks.

shadeofblue replied:
https://github.com/golemfactory/yapapi/issues/619

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+@dataclass
+class Node:
+    network: "Network"
+    node_id: str
```
add docstrings

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+    node_id: str
+    ip: str
+
+    def get_deploy_args(self) -> Dict:
```
add docstring

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+            )
+
+    async def add_address(self, ip: str):
+        """Assign the given IP address to the requestor in the Network."""
```
add :params:

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+        await self._net_api.add_address(self, ip)
+
+    async def add_node(self, node_id: str, ip: Optional[str] = None) -> Node:
+        """Add a new node to the network."""
```
add :params:

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+        await network.add_address(network.owner_ip)
+        return network
+
+    def __init__(
```
docstring

johny-b commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+        """Add a new node to the network."""
+        if ip:
+            self._ensure_ip_in_network(ip)
+            if ip in self.nodes_dict.keys():
```
I think `keys()` is not pythonic?

shadeofblue replied:
ah, good point

johny-b commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+        await self._net_api.add_node(self, node_id, ip)
+
+        node = Node(network=self, node_id=node_id, ip=ip)
+        self._nodes[node_id] = node
```
I'm not sure, but I think we could have a race condition here?

IP is added to a list of "used" ips here, in line 145.
If we have two calls to `network.add_node` running at the same time with **the same IP** they could pass the validation in line 134 (`if "ip" in self.nodes_dict.keys()`) and both get to `await self._net_api.add_node(...)` and I don't think we want this.

shadeofblue replied:
addressed this with a lock

johny-b commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+
+    def _next_address(self) -> IpAddress:
+        try:
+            return next(self._hosts)  # type: ignore
```
I'd prefer to `str()` the address here. We never want a non-str host anywhere.

johny-b commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+                f"The given IP ('{ip}') address must belong to the network ('{self._ip_network.with_netmask}')."
+            )
+
+    async def add_address(self, ip: str):
```
I would rename this to `add_owner_address`, or `add_requestor_address`.

shadeofblue replied:
done

johny-b commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+        network_id: Optional[str] = None,
+        mask: Optional[str] = None,
+        gateway: Optional[str] = None,
+    ):
```
A docstring with :params: would be useful here.

johny-b reviewed (changes requested):
There's a bunch of comments.

Nothing very important, except one possible race condition, but I think we should make maximum effort to make this perfect, it will stay with us forever : )

Also: everytime I `ctrl+C` the `example/ssh.py` script I get:
```
Task was destroyed but it is pending!
task: <Task pending name='Task-42' coro=<Queue.get() running at /usr/lib/python3.8/asyncio/queues.py:163> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fad3dc91760>()]>>
```

shadeofblue commented on examples/ssh/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+RUN echo "UseDNS no" >> /etc/ssh/sshd_config && \
+    echo "PermitRootLogin yes" >> /etc/ssh/sshd_config && \
+    echo "PasswordAuthentication yes" >> /etc/ssh/sshd_config
+RUN echo -e "{pass}.8\n{pass}.8" | passwd
```
sure, we can change it ;) ... probably Marek thought `dupa.8` would be a bit too casual ;)))

shadeofblue replied:
changed

shadeofblue commented on tests/factories/net.py:
```diff
@@ -0,0 +1,21 @@
+import faker
+from unittest import mock
+
+from yapapi.net import Network
```
okay, made (almost) all of them `network.py` except the rest API binding, since the low level binding module is called `ya_net`

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+
+    @property
+    def owner_ip(self) -> str:
+        return str(self._owner_ip)
```
ah, my bad, the original type was mismatched and I have now corrected it to be an IpAddress

shadeofblue commented on examples/ssh/ssh.py:
```diff
@@ -0,0 +1,142 @@
+        print(
+            f"""Connect with:
+{TEXT_COLOR_CYAN}
+ssh -o ProxyCommand='websocat asyncstdio: ws://127.0.0.1:7465/net-api/v1/net/{net}/tcp/{ip}/22 --binary -H=Authorization:"Bearer {app_key}"' root@{ip}
```
hmm, good point indeed

shadeofblue replied:
updated

shadeofblue commented on yapapi/ctx.py:
```diff
@@ -337,6 +343,11 @@ def provider_name(self) -> Optional[str]:
+    @property
+    def provider_id(self) -> str:
+        """Return the id of the provider associated with this work context."""
+        return self._agreement_details.raw_details.offer.provider_id  # type: ignore
```
ah, right... only it can't ... since the `provider_id` there is not one of the offer properties which we use to extract a Model ...

shadeofblue replied:
added an explanation

shadeofblue commented on tests/factories/net.py:
```diff
@@ -0,0 +1,21 @@
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
+        pool = futures.ThreadPoolExecutor()
+        return pool.submit(asyncio.run, model_class.create(*args, **kwargs)).result()
```
added

shadeofblue commented on yapapi/golem.py:
```diff
@@ -211,13 +223,42 @@ async def main():
+        :param gateway: Optional gateway address for the network
+        :return: a Network object allowing further manipulation of the created VPN
+        """
+        async with self._root_api_session.get(f"{self._api_config.root_url}/me") as resp:
```
hmm... I'd rather have this call specifically here, just like we don't provide wrappers for other API calls in the Engine itself

shadeofblue commented on yapapi/engine.py:
```diff
@@ -259,6 +260,16 @@ def report_shutdown(*exc_info):
+            net_client = await stack.enter_async_context(self._api_config.net())
+            self._net_api = rest.Net(net_client)
+
+            # TODO replace with a proper REST API client once ya-client and ya-aioclient are updated
```
ok, added an issue and yes, I'd leave the root api session here

shadeofblue commented on yapapi/golem.py:
```diff
@@ -153,6 +157,14 @@ async def run_service(
+        :param network: optional Network, representing a VPN to attach this Cluster's instances to
+        :param network_addresses: optional list of addresses to assign to consecutive spawned instances.
+            If not provided, the addresses will be assigned automaticaly.
+            Otherwise, the number of addresses should correspond with the number of instances determined
```
actually, I think it, "correspond *with*" is a slightly better choice here, at least according to: https://english.stackexchange.com/questions/142495/correspond-to-vs-correspond-with

johny-b replied:
> actually, I think it, "correspond _with_" is a slightly better choice here, at least according to: https://english.stackexchange.com/questions/142495/correspond-to-vs-correspond-with

I've read the same and thought `or agree **in amount**, position, etc.` is crucial here.

shadeofblue commented on yapapi/golem.py:
```diff
@@ -211,13 +223,42 @@ async def main():
+
+        :param ip: the IP address of the network. May contain netmask, e.g. "192.168.0.0/24"
+        :param owner_ip: the desired IP address of the requestor node within the newly-created Network
+        :param mask: Optional netmask (only if not provided within the `ip` argument)
```
hmm... I'd leave it as it is - I imagine people who prefer to define the networks with a netmask with the IP address and those who wish to provide it explicitly with a separate argument

shadeofblue commented on yapapi/services.py:
```diff
@@ -138,18 +140,33 @@ def cluster(self) -> "Cluster":
         return self._ctx.provider_name
 
+    @property
+    def provider_id(self) -> str:
```
it's because the underlying properties have those types

johny-b replied:
I don't think this explains anything?

If there is any reason (like a macro-scale reason) to have a Service with defined `provider_id`, but without `provider_name`, this could be documented. If there is no real reason, but this is only an implementation detail, we might note this as a possible candidate for some refactoring. I understand this might not be a part of this PR, but this **is** some sort of a problem.

shadeofblue commented on yapapi/services.py:
```diff
@@ -779,6 +821,10 @@ def spawn_instances(
+        :param network_addresses: optional list of network addresses in case the Cluster is
+            attached to VPN. If the list is not provided (or if the number of elements is less
+            than the number of spawned instances), any instances for which the addresses have not
+            been given, will be assigned an address automatically.
```
okay, I don't think it makes sense to resolve this issue in this pull request ... plus, there _is_ a reason why it must be part of "spawn_instances" -> because `spawn_instances` can be called on the `Cluster` itself too

johny-b replied:
I don't think this is a good reason. Why `Cluster.spawn_instances` would need to have exactly the same interface as `Golem.run_service`? They seem to be totally different things, despite first being called by the second.

But I agree this is not a part of this PR.

shadeofblue replied:
they don't _need_ to have exactly the same interface ... but if they (partially) do the same thing, the interface for those pieces should stay consistent

shadeofblue commented on yapapi/rest/configuration.py:
```diff
@@ -119,6 +122,7 @@ def resolve_url(given_url: Optional[str], env_val: str, prefix: str) -> str:
         self.__activity_url: str = resolve_url(
             activity_url, "YAGNA_ACTIVITY_URL", "/activity-api/v1"
         )
+        self.__net_url: str = resolve_url(net_url, "YAGNA_NET_URL", "/net-api/v1")
```
because that's the name of the module...

johny-b replied:
What module?
Sorry, I don't know which names we are allowed to change and which names are fixed.

shadeofblue replied:
`ya_net` ... it's not fixed -> just I prefer for that to be analogous

shadeofblue commented on yapapi/net.py:
```diff
@@ -0,0 +1,156 @@
+                    "id": self.network.network_id,
+                    "ip": str(self.network.network_address),
+                    "mask": str(self.network.netmask),
+                    "nodeIp": str(self.ip),
```
they're not

johny-b replied:
Which exactly is not?

`self.ip`?
```
     ip: str
```

`network.network_address`?
```
    @property
    def network_address(self) -> str:
        """The network address of this network, without a netmask."""
        return str(self._ip_network.network_address)
```

or `self.network.netmask`?
```
    @property
    def netmask(self) -> str:
        """The netmask of this network."""
        return str(self._ip_network.netmask)
```

shadeofblue replied:
okay, maybe they weren't originally... I rest my case ...

shadeofblue replied:
updated

shadeofblue commented on yapapi/rest/net.py:
```diff
@@ -0,0 +1,31 @@
+    models as yan,
+)
+
+from yapapi.net import Network
```
yeah, it's an artifact of the refactoring process -> I'll update it indeed

shadeofblue replied:
updated

shadeofblue commented on yapapi/rest/configuration.py:
```diff
@@ -140,6 +144,16 @@ def activity_url(self) -> str:
+        return self.__net_url
+
+    @property
+    def root_url(self) -> str:
```
I don't know ... I prefer to have `root` here

shadeofblue commented on examples/ssh/ssh.py:
```diff
@@ -0,0 +1,142 @@
+        self._ctx.run("/bin/bash", "-c", "ssh-keygen -A")
+        self._ctx.run("/bin/bash", "-c", "/usr/sbin/sshd")
+        self._ctx.run("/bin/bash", "-c", "sleep 3600")
+        yield self._ctx.commit(timeout=timedelta(minutes=30))
```
updated too ... but now we have two cancelled tasks ;)

shadeofblue commented:
> There's a bunch of comments.
> 
> Nothing very important, except one possible race condition, but I think we should make maximum effort to make this perfect, it will stay with us forever : )
> 
> Also: everytime I `ctrl+C` the `example/ssh.py` script I get:
> 
> ```
> Task was destroyed but it is pending!
> task: <Task pending name='Task-42' coro=<Queue.get() running at /usr/lib/python3.8/asyncio/queues.py:163> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fad3dc91760>()]>>
> ```

okay, addressed almost all of the comments apart from those which I don't think are relevant here or those that I don't agree with ... 

as for the Ctrl-C, I added a cluster shutdown but it still blows if there's just one provider because of https://github.com/golemfactory/yapapi/issues/544

johny-b commented on yapapi/network.py:
```diff
@@ -0,0 +1,230 @@
+        self._nodes: Dict[str, Node] = dict()
+        """the mapping between a Golem node id and a Node in this VPN."""
+
+        self._nodes_lock = asyncio.Lock()
```
I'm not sure, but I think this lock is redundant now?

There are no `await`s in the locked sections (in both places where it is used), so this code will either way execute as a single part.

The important part that changed since the previous version is 
```
self._nodes[node_id] = node
```
being moved **before** the only await in this function (`await self._net_api.add_node(...)`).

In other words, as long as we added a `node` to `self._nodes` after the `await`, the lock was necessary. Now it's not.

shadeofblue replied:
I'd leave it be so that we don't need to think about this if, e.g. a future change adds some async code inside

johny-b reviewed (approved):
Very nice, thank you for addressing all the comments.

I think there is still one thing that *should* change, but I might be wrong, and even if I'm right it's more-or-less harmless, so there's a well deserved :+1: : )
